### PR TITLE
Transcripts - Support caching

### PR DESCRIPTION
**Pull request by ashiagr**
## Description
This supports caching for transcripts.

#### Test.1 For episode that is not downloaded

1. Launch the app
2. Play an episode (A) with transcript and load transcript
3. Play another episode (B) with transcript and load transcript
4. Disable internet
5. Play previous episode (A) and try to load transcript
6. ✅ Notice that the transcript is loaded from cache
7.  Enable internet

#### Test.2 For downloaded episode

1. Launch the app
2. Download an episode (C) with transcript
3. Disable internet
5. Play the downloaded episode and try to load the transcript
6. ✅ Notice that the transcript is loaded from cache

## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [ ] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [ ] Any jetpack compose components I added or changed are covered by compose previews
- [ ] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.
 
#### I have tested any UI changes...

- [ ] with different themes
- [ ] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

### CODE REVIEW

File: modules/services/servers/src/main/java/au/com/shiftyjelly/pocketcasts/servers/di/ServersModule.kt [hunk 1]
```diff
@@ -107,7 +107,7 @@ class ServersModule {
         }
 
         fun provideCache(folder: String, context: Context): Cache {
-            val cacheSize = 10 * 1024 * 1024 // 10 MB
+            val cacheSize = 50 * 1024 * 1024 // 50 MB
```
Comment (ashiagr) [inline modules/services/servers/src/main/java/au/com/shiftyjelly/pocketcasts/servers/di/ServersModule.kt:110]: Reusing exiting `HttpCache`, increased cache size to allow transcripts caching.

File: modules/services/servers/src/main/java/au/com/shiftyjelly/pocketcasts/servers/podcast/PodcastCacheServer.kt [hunk 2]
```diff
@@ -80,6 +80,10 @@ interface PodcastCacheServer {
     @GET
     suspend fun getShowNotesChapters(@Url url: String): RawChaptersResponse
 
+    @GET
+    @Headers("Cache-Control: max-stale=7776000") // Use cache available for 90 days, if not cached use network
+    suspend fun getTranscript(@Url url: String): Response<ResponseBody>
```
Comment (ashiagr) [inline modules/services/servers/src/main/java/au/com/shiftyjelly/pocketcasts/servers/podcast/PodcastCacheServer.kt:85]: OkHttp automatically applies `Etag`, `Cache-Control`, etc logic ([ref](https://futurestud.io/tutorials/retrofit-2-activate-response-caching-etag-last-modified#:~:text=OkHttp%20will%20automatically%20apply%20Etag%2C%20Cache%2DControl%2C%20etc%20logic%20on%20every%20request%20for%20you)), I'm just overriding the max-stale value.

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/shownotes/ShowNotesManager.kt [hunk 3]
```diff
@@ -27,7 +28,7 @@ class ShowNotesManager @Inject constructor(
     suspend fun downloadToCacheShowNotes(podcastUuid: String, episodeUuid: String) {
         serverShowNotesManager.downloadToCacheShowNotes(
             podcastUuid = podcastUuid,
-            processShowNotes = { showNotesProcessor.process(episodeUuid, it) },
+            processShowNotes = { showNotesProcessor.process(episodeUuid, it, LoadTranscriptSource.DOWNLOAD_EPISODE) },
```
Comment (ashiagr) [inline modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/shownotes/ShowNotesManager.kt:31]: When source is `DOWNLOAD_EPISODE`, transcript is also downloaded.

**Comment by MiSikora** [comment]
Transcripts don't load for me when I go offline. Am I missing some step?

https://github.com/user-attachments/assets/19973a32-f200-4f24-82d6-96672a6eecb6

**Comment by ashiagr** [comment]
It was tricky to solve. 

Server response for the episode you tested included `cache-control: max-age=0, private, must-revalidate`. Our interceptor `INTERCEPTOR_CACHE_MODIFIER` could not modify the cache-control headers while caching the network response. It is recommended to [use a network interceptor](https://stackoverflow.com/a/70052313/193545) for this purpose. This is done in d43808933e62bd745de67cdc092aa5a39533b04f.

I load transcript from cache or server as follows: 

- if `forceRefresh=false`, load from cache
    - if cache response failed, and network available, load from network
    - if cache response failed, and no network, return no network exception
- if `forceRefresh=true`, 
   - and network available, load from network
   - and no network, load from cache

`forceRefresh` is true when episode is downloaded or the transcript is pulled-to-refresh.

Can you give it another try?

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/podcast/TranscriptsManagerImpl.kt [hunk 4]
```diff
@@ -33,10 +48,50 @@ class TranscriptsManagerImpl @Inject constructor(
...
 enum class TranscriptFormat(val mimeType: String) {
     SRT("application/srt"),
     VTT("text/vtt"),
     HTML("text/html"),
 }
+
+enum class LoadTranscriptSource {
+    DOWNLOAD_EPISODE, // When transcript is downloaded as part of episode download
+    DEFAULT,
+}
```
Comment (MiSikora) [inline modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/podcast/TranscriptsManagerImpl.kt:97]: I think this should be rather defined next to the interface and not next to the implementation.

**Review by MiSikora — APPROVED** [review]
It works as expected. I'm not sure about using OkHttp's cache for caching transcripts. It might lead to somewhat poor UX because the cache will be populated by other things like network requests, images, etc. I think it would be better to have a separate cache just for transcripts.

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/podcast/TranscriptsManagerImpl.kt (hunk 4 above)
Comment (ashiagr) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/podcast/TranscriptsManagerImpl.kt]: Moved to interface in e13252bd493f70b6dfb038df9cbf0d35b922d1aa

**Comment by ashiagr** [comment]
> I think it would be better to have a separate cache directory that we manage just for transcripts.

I separated out transcript cache in [3258b88](https://github.com/Automattic/pocket-casts-android/pull/2570/commits/3258b883fa260f998fca11d12708b741b2dd9f7b). So podcast server caching remains unchanged and unaffected by these changes.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**